servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ScanWithLifetimePublisher.java

```diff
@@ -0,0 +1,254 @@
+         * This state helps differentiate between cancel and reentrant completion.
+         */
+        private static final int STATE_FINALIZE_AFTER_REENTRY = 4;
+        /**
+         * Special state to handle the case where a reentry to a terminal (eg. onComplete) signal needs to defer
+         * a pending finalization to the root frame onNext.
+         */
+        private static final int STATE_FINALIZE_PENDING_AFTER_REENTRY = 5;
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<ScanWithLifetimeSubscriber> stateUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(ScanWithLifetimeSubscriber.class, "state");
+
+        private volatile int state = STATE_UNLOCKED;
+
+        private final ScanWithLifetimeMapper<? super T, ? extends R> mapper;
+
+        ScanWithLifetimeSubscriber(final Subscriber<? super R> subscriber,
+                           final ScanWithLifetimeMapper<? super T, ? extends R> mapper,
+                           final SignalOffloader signalOffloader, final AsyncContextMap contextMap,
+                           final AsyncContextProvider contextProvider) {
+            super(subscriber, mapper, signalOffloader, contextMap, contextProvider);
+            this.mapper = mapper;
+        }
+
+        @Override
+        protected void onCancel() {
+            // This is not a serial invocation.
+            // Finalize only when CAS succeeds and previous state was UNLOCKED or FINALIZE_PENDING, otherwise:
+            // - If BUSY the signal will do the finalization
+            // - If FINALIZED or FINALIZED_REENTRY nothing needs to be done
+            final int prevState = stateUpdater.getAndSet(this, STATE_FINALIZED);
+            if (prevState == STATE_UNLOCKED || prevState == STATE_FINALIZE_PENDING) {
+                finalize0();
+            }
+        }
+
+        @Override
+        public void onNext(@Nullable final T t) {
+            final int prevState = state;
+            if (prevState == STATE_FINALIZED) {
+                return;
+            }
+            final boolean reentry = prevState == STATE_BUSY;
+
+            // No need to loop since the signals are guaranteed to be signalled serially,
+            // thus a CAS fail means cancellation won, so we can safely ignore it.
+            if (!stateUpdater.compareAndSet(ScanWithLifetimeSubscriber.this, prevState, STATE_BUSY)) {
+                return;
+            }
+
+            try {
+                super.onNext(t);
+            } finally {
+                // Re-entry -> don't unlock
+                if (!reentry) {
+                    if (state == STATE_FINALIZE_PENDING_AFTER_REENTRY) {
+                        // If a re-entrant terminal didn't complete due to missing demand, we take over and set
+                        // state to STATE_FINALIZE_PENDING.
+                        if (!stateUpdater.compareAndSet(this, STATE_FINALIZE_PENDING_AFTER_REENTRY,
```

Comment:
if `onCancel` happens concurrently we may double-invoke `finalize()`. we should be more explicit about state transitions.

I extended your work here and to demo explicit state transitions which avoid concurrent finalization https://github.com/scottmitch/servicetalk-1/tree/scan_with_lifetime_demo.

Reply:
FWIW there is no double `finalize()` involved here, double checked the code. Let me know if you spend more time on it and figure out a path that leads to that behavior, would be nice to know if I missed sth.
Regardless, after our offline chat, we agreed to bring in your changes which demonstrate better clarity. They do have extra CAS in some cases but that can be an optimization we can look for later.